### ISSUE TITLE
Bad typer complexity (?) when checking that types are not directly recursive 

### ISSUE THREAD

**Issue by yakobowski** [ocaml/ocaml#11150]
We are hitting what seems to be a problem in the typer (complexity-wise) when "closing" recursive types, i.e. when doing the following operation

```
type 'expr expr = [ ... | `Unop of unop *  'expr ]
type expr' = expr' expr  (* "closing" the type *)
```

The slow operation is the final one. Investigations revealed that nearly all the time is spent in the function `Typedecl.check_well_founded`. Closing types as above does not scale at all with multiple recursive types that themselves have multiple type variables. We have reproduced that on automatically generated code (specifically crafted for this issue), but the initial problem we encountered was on manually written code, which was of a completely reasonable size. At the peak of our experiments, compilation of our main file (4 kLoC, 7 kLoC after PPX expansion) was taking 2.5 minutes, nearly all of them spent closing our main types. This made the edit-compile cycle very tedious, especially since Merlin is completely dead in the water on such code.

The issue can be reproduced from the code generated by compiling [gen.ml](https://github.com/ocaml/ocaml/files/8389166/gen.ml.txt)  and running it with arguments `4 4`, `5 5` and `6 6`. More precisely, the following lines give interesting results on my machine.
```
$ ocamlc gen.ml -o gen
$ ./gen 4 4  > repro.ml  && /usr/bin/time ocamlc -c repro.ml 
0.65user 0.00system 0:00.66elapsed 99%CPU (0avgtext+0avgdata 14228maxresident)k
0inputs+16outputs (0major+1642minor)pagefaults 0swaps
$ ./gen 5 5  > repro.ml  && /usr/bin/time ocamlc -c repro.ml
68.86user 0.25system 1:09.25elapsed 99%CPU (0avgtext+0avgdata 14344maxresident)k
0inputs+16outputs (0major+1916minor)pagefaults 0swaps
$ ./gen 6 6  > repro.ml  && /usr/bin/time ocamlc -c repro.ml 
11556.35user 0.42system 3:12:40elapsed 99%CPU (0avgtext+0avgdata 15676maxresident)k
0inputs+24outputs (0major+2074minor)pagefaults 0swaps
```

So 0.7s, 69s and 3h 12 mins, even though at each step we only add one new type. For reference, [repro.ml](https://github.com/ocaml/ocaml/files/8389153/repro.ml.txt) with n=6 is only 64 lines long.

The problem is very impactful for us, and we would like to help fixing it. However, given the very limited understanding of the OCaml typer in our team, I don't really know what to start looking for. Would you have any advice?

------------------

A remark that may be related. In the `repro.ml` file, all the types are mutually recursive. However, consider the variant below, generated by passing the argument `Rectangle` instead of `Square` in `gen.ml`.

```
type ('t1) t1 = [
  | `C1_1 of 't1
  | `C1_2 of 't1
  | `C1_3 of 't1
  | `C1_4 of 't1k
]
and ('t1, 't2) t2 = [
  | `C2_1 of 't1 * 't2
  | `C2_2 of 't1 * 't2
  | `C2_3 of 't1 * 't2
  | `C2_4 of 't1 * 't2
]
and ('t1, 't2, 't3) t3 = [
  | `C3_1 of 't1 * 't2 * 't3
  | `C3_2 of 't1 * 't2 * 't3
  | `C3_3 of 't1 * 't2 * 't3
  | `C3_4 of 't1 * 't2 * 't3
]
and ('t1, 't2, 't3, 't4) t4 = [
  | `C4_1 of 't1 * 't2 * 't3 * 't4
  | `C4_2 of 't1 * 't2 * 't3 * 't4
  | `C4_3 of 't1 * 't2 * 't3 * 't4
  | `C4_4 of 't1 * 't2 * 't3 * 't4
]

type 'a typed = {node: 'a; loc: int}

type ('t1) t1_typed = ('t1) t1 typed
type ('t1, 't2) t2_typed = ('t1, 't2) t2 typed
type ('t1, 't2, 't3) t3_typed = ('t1, 't2, 't3) t3 typed
type ('t1, 't2, 't3, 't4) t4_typed = ('t1, 't2, 't3, 't4) t4 typed

type t1' = (t1') t1_typed
and t2' = (t1', t2') t2_typed
and t3' = (t1', t2', t3') t3_typed
and t4' = (t1', t2', t3', t4') t4_typed
```

With N sufficiently big (here the generation was done with arguments 4 4), compiling this program starts taking a lot of time. However, observe that the closing operation can be rewritten into
```
type t1' = (t1') t1_typed
type t2' = (t1', t2') t2_typed
type t3' = (t1', t2', t3') t3_typed
type t4' = (t1', t2', t3', t4') t4_typed
```
This makes the typing phase instantaneous. This rewriting can of course be done by the user, but it may be a bit complex because of name capture. So it would be really interesting to have the compiler perform this reorganization internally.

-------------------

Finally a bit of context. The reason we are using such "open" types is that we are writing an Ada -> Infer frontend based on a [Nanopasses](https://nanopass.org/) approach. Except that this is in OCaml, so each type and each transformation is fully typed. We are keeping the types open for the following reason. Assume the four types below

```
type 'expr expr1 = [ ... | `C of 'expr | `D of 'expr ]
type expr1 = expr1' expr1
type 'expr expr2 = [ ... | `C of 'expr | `E of 'expr]
type expr2' = expr2' expr2
```
A translation function from `expr1'` to `expr2'` would be decomposed into two parts:
- a function that takes a `expr2' expr1` (all but the first level of constructors have been translated), and return a `expr2' expr2` (i.e. a `expr2'`). In particular, this function has to remove the constructor `D ...`.
- a function that recurses within the structure of `expr1`, and propagates the translation.

We need the open versions of the types to annotate the boilerplate functions, and the closed versions to check that our transformation functions are fully well-typed. (This is actually a bit of an over-simplification, but roughly conveys the idea.)

**Comment by yakobowski** [comment]
> The problem is very impactful for us, and we would like to help fixing it. However, given the very limited understanding of the OCaml typer in our team, I don't really know what to start looking for. Would you have any advice?

@garrigue Would you be able to suggest somewhere I could start looking at? Basically, anything that could help me understand what `Typedecl.check_well_founded` does, and how, would help. Thanks!

**Comment by gasche** [comment]
Some thoughts:
- I believe that `check_well_founded` is meant to forbid definitions such as `type 'a t = 'a * ('a t) option` unles `-rectypes` is passed.
- The code is hard to understand (it contains a creative mix of recursion, global state update, re-raised exceptions etc.) and not documented.
- The algorithm is a graph traversal to detect cycles; the cycle-detection criterion is that the encountered node is already present on the "path" already traversed from the root.
- In fact, some forms of cycles are safe, when they are guarded by an object or a variant; on encountering such a safe type constructor, the `parents` path is replaced by an empty set; in other words, the algorithm fails when it encounters a node that was encountered after the last "safe" type constructor.
- On encountering a `Tconstr(p, args)` node (as in `('a t) option`) that is not already an unsafe parent, the logic appears to be as follows: 
 1. If the type constructor `p` is part of the mutually-recursive definition, first check if one of the arguments contains an unsafe parent (and fail), or expand the definition of `p` and check the result.
 2. If the type constructor `p` is not part of the mutually-recursive definition, first check if one of the arguments contains an unsafe cycle; if it doesn't, succeed, otherwise expand the definition of `p` and check the result.


In your `repro.ml`, we are talking about the recursive definition of the closed types `t1', t2', ...`. When checking `t1'`, at the beginning just `t1'` is part of the unsafe/visited/parents nodes (not `t2'`, etc.). `t1_typed` is not part of the mutually-recursive definition, so first the arguments are checked and `t1'` (or any other) fails, so according to (2) above `(t1',...) t1_typed` is expanded into `(t1',...) t1 typed`. Then following the same logic, `(t1',...) t1` is checked, which in turns expands `t1`, which checks each constructor `` `C1_i of t1' * t2' * t3' * ...`` in turn. We passed under a safe constructor, so the `parents` set is empty, and so checking `t1'` succeeds immediately, but `t2'` needs to be expanded.

In other words: with the current algorithm, checking `t1'` will expand `t2'` once for each occurrence of `'t2` in the defintions of the `t1, t2, ..., t6`, and each check will then expand `t3`, then `t4`, then `t5`, then `t6` in some search paths (and make the same expansions in other orders in other search paths). The algorithm seems super-exponential (I think maybe `6 ^ (6!)`?).

I think that for your type-declaration pattern to work well, you need a less naive cyclic-graph-declaration aglorithm.

**Comment by yakobowski** [comment]
Thanks a lot Gabriel! This is already very helpful. I've started having a deeper look, and I have one last question: did you understand what the `ty0` argument of `check` is? It almost look like as if it was made for reporting only, but not quite. And it complicates rewriting the code with something that makes heavier use of memoization.

**Comment by gasche** [comment]
I think that `ty0` is indeed only used for error reporting. At the beginning it is the type expression whose declaration is being checked (`'a t` in my example I think?), and it gets updated when we decide to expand a definition because it is part of the same recursive group. In other words, it is the member of the mutual group that is nearest above the cycle found.

(It's not obvious from the `ty0` update code that it is only updated when `to_check p` holds, and in fact I am not sure that it is the case, but I believe that at least this is the *intent* of the code, as seen in the simpler previous version mentioned in the following.)

Some other remarks:
- In my analysis of the blowup behavior above, I neglected the fact that some memoization happens with `!visited`. The details are rather tricky because the memoization map is sometimes rolled back (when checking direct parameters without expansion), so I'm not sure what about of memoization you would get in your example, but the real complexity could be (sensibly) lower than what I said. It does appear to blow up, so it's certainly bad, but my explanation was simplistic.
- I found that the implementation was made significantly more complex in commit 848c20f27d1e92df359ef02987cc2dc7d4c54cc7 in 2016; the implementation before that feels simpler to me, so maybe you want to look at it instead: https://github.com/ocaml/ocaml/blob/848c20f27d1e92df359ef02987cc2dc7d4c54cc7~1/typing/typedecl.ml#L493-L539
- in the new implementation, there is a line that I don't understand
  https://github.com/ocaml/ocaml/blob/146766c352f02967cb98de6f57477618ff5f7c6f/typing/typedecl.ml#L683
  which appears to re-run a check on the constructor arguments, duplicating work that was done in the `arg_exn` computation. This may be a bug (or maybe not), and in particular it may cause a blowup in some unintended cases.

**Comment by bluddy** [comment]
Peeking in here, I read some of this code several years back and it's just so painful. Can we please add a task to insert copious comments in this typechecking stuff?

**Comment by gasche** [comment]
"The type-checker code could be clearer" is a meta-issue that does not need to be created, I think. We've worked on specific instances of that (see for example #9030), but it's work -- at least a full day of work. If the discussion/work goes in the direction that we understand enough to prepare a clarification PR, we'll do it. In the meantime anyone should feel free to do it if they want.

**Comment by garrigue** [comment]
> * in the new implementation, there is a line that I don't understand
>   https://github.com/ocaml/ocaml/blob/146766c352f02967cb98de6f57477618ff5f7c6f/typing/typedecl.ml#L683
>   
>   which appears to re-run a check on the constructor arguments, duplicating work that was done in the `arg_exn` computation. This may be a bug (or maybe not), and in particular it may cause a blowup in some unintended cases.

The second check is different, as `parents` is replaced by `Typeset.empty`.
I don't remember all the details, but one tries to detect illegal cycles, including local ones.
The code for `arg_exn` assumes that `p` may not be contractive, and as such calls `check ty0 parents` on parameters, except if `rec_ok`, which in this case is very restrictive (requires `-rectypes`, and applies only to datatypes).
In the case `to_check p`, we choose to follow this assumption, to fix #6505.
Otherwise, we will just expand `ty`, but we still want to check the well-roundedness of its parameters individually.

It is true that the check is superfluous if `rec_ok = true || Typeset.is_empty parents`.
I'm not sure this will change the complexity, but this may be worth adding.

**Comment by garrigue** [comment]
Looking again at the code, I think that it could separate better its two roles: discover recursion between definitions, and discover local cycle inside individual definitions. In particular, now that we assume that abbreviations inside the recursion are non-contractive, there seems to be no need to check deep inside them when they are expanded. This is probably the reason for the wrong complexity here.

**Comment by garrigue** [comment]
PR #11207 seems to solve the complexity problem.
It is careful not to duplicate work.
However there also exists situations where the current algorithm is more efficient; if we want to solve those two, we probably need to keep contractiveness in the environment too.

**Comment by yakobowski** [comment]
Thanks a lot @garrigue, much appreciated -- especially since my own attempts did not get very far :sweat_smile: 

I backported the patch to the OCaml compiler we use, and I can confirm the problem is entirely solved on our side:

Before
```
yakobowski@defiant:~/standalone-nanopass$ time dune build                    
real	3m41,597s
user	3m51,786s
sys	0m1,265s
```

After
```
yakobowski@defiant:~/standalone-nanopass$ time dune build
real	0m9,206s
user	0m12,195s
sys	0m1,220s
```

(This is on the "worst" version of our code. Since then we had found ways to improve the compilation times.)

**Comment by gasche** [comment]
@yakobowski the fix for this issue is now merged, it should be part of OCaml 5.1.
